packages/sitecore-jss-nextjs/src/services/component-props-service.ts

```diff
@@ -0,0 +1,176 @@
+
+      return module?.getServerSideProps;
+    };
+
+    return this.fetchComponentProps<GetServerSidePropsContext>(
+      fetchFunctionFactory,
+      layoutData,
+      context
+    );
+  }
+
+	/**
+	 * SSG mode
+	 * Fetch component props using getStaticProps function
+	 */
+  async fetchStaticComponentProps(
+    params: FetchComponentPropsArguments<GetStaticPropsContext>
+  ): Promise<ComponentProps> {
+    const { componentModule, layoutData, context } = params;
+
+    const fetchFunctionFactory = (componentName: string) => {
+      const module = componentModule(componentName);
+
+      return module?.getStaticProps;
+    };
+
+    return this.fetchComponentProps<GetStaticPropsContext>(
+      fetchFunctionFactory,
+      layoutData,
+      context
+    );
+  }
+
+  /**
+   * Traverse Layout Service data tree and call side effects on component level.
+   * Side effect function can be: getStaticProps (SSG) or getServerSideProps (SSR)
+   */
+  async fetchComponentProps<NextContext>(
+    fetchFunctionFactory: FetchFunctionFactory<NextContext>,
+    layoutData: LayoutServiceData,
+    context: NextContext
+  ): Promise<ComponentProps> {
+    // Array of side effect functions
+    const requests = this.collectRequests({
+      placeholders: layoutData.sitecore.route.placeholders,
+      fetchFunctionFactory,
+      layoutData,
+      context,
+    });
+
+    return await this.execRequests(requests);
+  }
+
+  /**
+   * Go through layout service data, check all renderings using displayName, which should make some side effects.
+   * Write result in requests variable
+   */
+  collectRequests<NextContext>(params: {
+    placeholders: PlaceholdersData;
+    fetchFunctionFactory: FetchFunctionFactory<NextContext>;
+    layoutData: LayoutServiceData;
+    context: NextContext;
+    requests?: ComponentPropsRequest<NextContext>[];
+  }) {
+    const { placeholders, fetchFunctionFactory, layoutData, context } = params;
+
+		// Will be called on first round
+    if (!params.requests) params.requests = [];
+
+    const renderings = this.flatRenderings(placeholders);
+
+    renderings.map((r) => {
+      const fetchFunc = fetchFunctionFactory(r.componentName);
+
+      if (fetchFunc) {
+        params.requests &&
+          params.requests.push({
+            fetch: fetchFunc,
+            rendering: r,
+            layoutData: layoutData,
+            context,
+          });
+      }
+
+      // If placeholders exist in current rendering
+      if (r.placeholders) {
+        this.collectRequests({
+          ...params,
+          placeholders: r.placeholders,
+        });
+      }
+    });
+
+    return params.requests;
+  }
+
+  /**
+   * Execute request for component props
+   */
+  async execRequests<NextContext>(requests: ComponentPropsRequest<NextContext>[]) {
+    const componentProps: ComponentProps = {};
+
+    const promises = requests.map((req) =>
+      req.fetch(req.rendering, req.layoutData, req.context).then((result) => {
```

Comment:
Should we add error handling for the individual fetch functions? It's my understanding that Promise.all would reject as soon as one promise rejects. So, if one of your components errors, the entire page would fail?

Reply:
Added `catch` handler, log error and write it inside component props.